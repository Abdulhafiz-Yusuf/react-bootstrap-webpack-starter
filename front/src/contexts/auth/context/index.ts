import { createContext } from 'react';

// #region types
export type User = {
  email: string;
};

export type AuthData = {
  isAuthenticated: boolean;
  isExpiredToken?: boolean;
  lastAuthDate?: Date;
  token?: string;
  user?: User;
};
// #endregion

// #region default context value
export const authDefault: AuthData = {
  isAuthenticated: false,
};
// #endregion

// #region context
const AuthContext = createContext<AuthData>({
  ...authDefault,
});

export const AuthContextProvider = AuthContext.Provider;
export const AuthContextConsumer = AuthContext.Consumer;
// #endregion
