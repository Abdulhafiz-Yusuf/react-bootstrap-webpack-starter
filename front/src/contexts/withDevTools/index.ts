// #region types
export type DevToolsMessageType = 'DISPATCH' | string;

export type DevToolsMessagePayload = {
  type?: string;
  state?: any;
};

export type DevToolsMessage = {
  type?: DevToolsMessageType;
  payload?: DevToolsMessagePayload;
};

export type DevTools = {
  connect: () => any;
  subscribe: (message: DevToolsMessage) => any;
  send: (action: { type: string; state?: any }, newState: any) => any;
  unsubscribe: () => any;
  dispatch: (action: { type: string }) => any;
  disconnect: () => any;
};
// #endregion

// #region constants
const isDEV = process.env.NODE_ENV === 'development';

export const withDevTools =
  isDEV &&
  typeof window !== 'undefined' &&
  (window as any).__REDUX_DEVTOOLS_EXTENSION__;

const devTools: DevTools = !withDevTools
  ? null
  : (window as any).__REDUX_DEVTOOLS_EXTENSION__.connect();
// #endregion

// #region devtools reducer
type State = {
  auth: any;
};
type Action = {
  type: string;
  state?: any;
};

const initialState: State = {
  auth: {},
};

export const reducer = (state: State = initialState, action: Action) => {
  /* eslint-disable no-unused-vars */
  switch (action.type) {
    // #region auth context
    case 'AUTH_CHECK_IS_AUTHENTICATED':
    case 'AUTH_CHECK_TOKEN_IS_EXPIRED':
    case 'AUTH_SET_TOKEN':
    case 'AUTH_SET_USER_INFO':
    case 'AUTH_DISCONNECT_USER': {
      const { type, state: context, ...rest } = action;
      return { ...state, user: { context, ...rest } };
    }
    // #endregion

    default:
      return state;
  }
  /* eslint-enable no-unused-vars */
};
// #endregion

// #region singleton devtools local state
let state: State;
// #endregion

// #region devToolsStore (redux like)

export const devToolsStore = !withDevTools
  ? null
  : {
      ...devTools,
      dispatch: (action: Action) => {
        // #region action validation
        if (!action) {
          throw new Error('devTools dispatched action should be defined');
        }
        if (typeof action === 'function') {
          throw new Error('devTools dispatched action should be an object');
        }
        if (typeof action !== 'object') {
          throw new Error('devTools dispatched action should be an object');
        }
        if (Array.isArray(action)) {
          throw new Error('devTools dispatched action should be an object');
        }
        // #endregion

        const newState = reducer(state, action);
        state = newState;

        devTools && devTools.send({ ...action }, newState);
      },
    };
// #endregion
