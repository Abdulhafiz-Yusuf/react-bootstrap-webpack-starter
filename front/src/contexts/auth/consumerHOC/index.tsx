import React, { Component } from 'react';
import wrapDisplayName from 'recompose/wrapDisplayName';
import { AuthContextConsumer } from '../context/index';
import { AuthProviderState } from '../providerComponent';

// #region flow types
export type AuthContextProps = {} & AuthProviderState;
// #endregion

// #region CONSUMER HOC
export default function withAuth(/* additionnal args if needed */) {
  return (BaseComponent: React.Component<any, any>) => {
    class WithAuth extends Component<any, any> {
      render() {
        const { ...passProps } = this.props;

        return (
          <AuthContextConsumer>
            {(fromAuthProps: AuthContextProps) => (
              <BaseComponent {...fromAuthProps} {...passProps} />
            )}
          </AuthContextConsumer>
        );
      }
    }

    // #region add static displayName for dev
    /* eslint-disable no-process-env */
    if (process.env.NODE_ENV !== 'production') {
      // @ts-ignore
      WithAuth.displayName = wrapDisplayName(BaseComponent, 'WithAuth');
    }
    /* eslint-enable no-process-env */
    // #endregion

    return WithAuth;
  };
}
// #endregion
