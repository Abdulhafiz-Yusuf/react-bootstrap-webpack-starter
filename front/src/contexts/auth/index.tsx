import React, { Component, createContext } from 'react';
import auth from '../../services/auth';
import { devToolsStore } from '../withDevTools';

// #region types
export type AuthData = {
  isAuthenticated: boolean;
  isExpiredToken?: boolean;
  lastAuthDate?: Date;
  token?: string;
  user?: User;
};

export type AuthProviderState = {
  checkIsAuthenticated: () => boolean;
  checkTokenIsExpired: () => boolean;
  setToken: (token: string) => any;
  setUserInfo: (user: User) => any;
  disconnectUser: () => boolean;
} & AuthData;

export type AuthProviderProps = {
  initialState: {} & AuthData;
};
// #endregion

// #region context
export const AuthContext = createContext<AuthProviderState | null>(null);
// #endregion

// #region constants
const initialState: AuthData = {
  token: '',
  user: undefined,
  isAuthenticated: false,
  isExpiredToken: true,
  lastAuthDate: undefined,
};
// #endregion

// #region PROVIDER component
export class AuthProvider extends Component<
  AuthProviderProps,
  AuthProviderState
> {
  static defaultProps = {
    initialState: {
      ...initialState,
    },
  };

  // #region lifecyle
  constructor(props: AuthProviderProps) {
    super(props);

    // initialize state in constructor (otherwise function won't be passed)
    this.state = {
      ...this.props.initialState,
      checkIsAuthenticated: this.checkIsAuthenticated,
      checkTokenIsExpired: this.checkTokenIsExpired,
      disconnectUser: this.disconnectUser,
      setToken: this.setToken,
      setUserInfo: this.setUserInfo,
    };
  }

  render() {
    const { children } = this.props;

    return (
      <AuthContext.Provider
        value={{
          ...this.state,
        }}
      >
        {children}
      </AuthContext.Provider>
    );
  }
  // #endregion

  checkIsAuthenticated = (): boolean => {
    const checkUserHasId = (user: User) => user.id;
    const user = auth.getUserInfo();
    const isAuthenticated = !!(auth.getToken() && checkUserHasId(user));

    devToolsStore &&
      devToolsStore.dispatch({
        type: 'AUTH_CHECK_IS_AUTHENTICATED',
        state: { ...this.state, isAuthenticated },
      });

    this.setState({ isAuthenticated });
    return isAuthenticated;
  };

  checkTokenIsExpired = (): boolean => {
    const token = auth.getToken();
    const isExpiredToken = auth.isExpiredToken(token);

    devToolsStore &&
      devToolsStore.dispatch({
        type: 'AUTH_CHECK_TOKEN_IS_EXPIRED',
        state: { ...this.state, isExpiredToken },
      });

    this.setState({ isExpiredToken });
    return isExpiredToken;
  };

  setToken = (token: string = '') => {
    auth.setToken(token);
    const isExpiredToken = auth.isExpiredToken(token);

    devToolsStore &&
      devToolsStore.dispatch({
        type: 'AUTH_SET_TOKEN',
        state: { ...this.state, token, isAuthenticated: true, isExpiredToken },
      });

    this.setState({ token, isAuthenticated: true, isExpiredToken });
  };

  setUserInfo = (user: User) => {
    if (typeof user === 'object') {
      auth.setUserInfo(user);

      devToolsStore &&
        devToolsStore.dispatch({
          type: 'AUTH_SET_USER_INFO',
          state: { ...this.state, user },
        });

      this.setState({ user });
    }
  };

  disconnectUser = (): boolean => {
    auth.clearAllAppStorage();
    this.checkIsAuthenticated();

    devToolsStore &&
      devToolsStore.dispatch({
        type: 'AUTH_DISCONNECT_USER',
        state: { ...initialState },
      });

    this.setState({ ...initialState });

    return true;
  };
}

export default AuthProvider;
// #endregion
