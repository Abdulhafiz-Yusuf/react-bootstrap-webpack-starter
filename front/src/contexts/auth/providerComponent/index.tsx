import React, { Component } from 'react';
import { AuthContextProvider, AuthData } from '../context';
import auth from '../../../services/auth';
import { devToolsStore } from '../../withDevTools';

// #region types
export type AuthProviderProps = {
  initialState: {} & AuthData;
};
// #endregion

// #region constants
const initialState: AuthData = {
  token: '',
  user: undefined,
  isAuthenticated: false,
  isExpiredToken: true,
  lastAuthDate: undefined,
};
// #endregion

// #region PROVIDER component
export default class AuthProvider extends Component<
  AuthProviderProps,
  AuthProviderState
> {
  static defaultProps = {
    initialState: {
      ...initialState,
    },
  };

  // #region lifecyle
  constructor(props: AuthProviderProps) {
    super(props);

    // initialize state in constructor (otherwise function won't be passed)
    this.state = {
      ...this.props.initialState,
      checkIsAuthenticated: this.checkIsAuthenticated,
      checkTokenIsExpired: this.checkTokenIsExpired,
      disconnectUser: this.disconnectUser,
      setToken: this.setToken,
      setUserInfo: this.setUserInfo,
    };
  }

  render() {
    const { children } = this.props;

    return (
      <AuthContextProvider
        value={{
          ...this.state,
        }}
      >
        {children}
      </AuthContextProvider>
    );
  }
  // #endregion

  checkIsAuthenticated = (): boolean => {
    const checkUserHasId = (user: User) => user.id;
    const user = auth.getUserInfo() ? auth.getUserInfo() : null;
    const isAuthenticated = !!(auth.getToken() && checkUserHasId(user));

    devToolsStore &&
      devToolsStore.dispatch({
        type: 'AUTH_CHECK_IS_AUTHENTICATED',
        state: { ...this.state, isAuthenticated },
      });

    this.setState({ isAuthenticated });
    return isAuthenticated;
  };

  checkTokenIsExpired = (): boolean => {
    const token = auth.getToken();
    const isExpiredToken = auth.isExpiredToken(token);

    devToolsStore &&
      devToolsStore.dispatch({
        type: 'AUTH_CHECK_TOKEN_IS_EXPIRED',
        state: { ...this.state, isExpiredToken },
      });

    this.setState({ isExpiredToken });
    return isExpiredToken;
  };

  setToken = (token: string = '') => {
    auth.setToken(token);

    devToolsStore &&
      devToolsStore.dispatch({
        type: 'AUTH_SET_TOKEN',
        state: { ...this.state, token, isAuthenticated: true },
      });

    this.setState({ token, isAuthenticated: true });
  };

  setUserInfo = (user: User) => {
    if (typeof user === 'object') {
      auth.setUserInfo(JSON.stringify(user));

      devToolsStore &&
        devToolsStore.dispatch({
          type: 'AUTH_SET_USER_INFO',
          state: { ...this.state, user },
        });

      this.setState({ user });
    }
  };

  disconnectUser = (): boolean => {
    auth.clearAllAppStorage();
    this.checkIsAuthenticated();

    devToolsStore &&
      devToolsStore.dispatch({
        type: 'AUTH_DISCONNECT_USER',
        state: { ...initialState },
      });

    this.setState({ ...initialState });

    return true;
  };
}
// #endregion
