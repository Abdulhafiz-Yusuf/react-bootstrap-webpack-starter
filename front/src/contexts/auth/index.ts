import withAuth from './consumerHOC';
import { AuthData } from './context';
import AuthProvider from './providerComponent';
import { AuthProviderState } from './providerComponent';

export default {
  withAuth,
  AuthData,
  AuthProvider,
  AuthProviderState,
};
