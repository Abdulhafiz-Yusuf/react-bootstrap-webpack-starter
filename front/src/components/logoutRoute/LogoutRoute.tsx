import React from 'react';
import { useEffect, useContext } from 'react';
import { Route, Redirect } from 'react-router-dom';
import { RouteComponentProps } from 'react-router';
import { AuthProviderState, AuthContext } from '../../contexts/auth';

type OwnProps = {};
type Props = OwnProps & RouteComponentProps;

function LogoutRoute(props: Props) {
  const auth = useContext<AuthProviderState | null>(AuthContext);

  useEffect(() => {
    auth && auth.disconnectUser();
  }, []);

  return (
    <Route {...props}>
      <Redirect to={{ pathname: '/login' }} />
    </Route>
  );
}

LogoutRoute.displayName = 'LogoutRoute';

export default LogoutRoute;
