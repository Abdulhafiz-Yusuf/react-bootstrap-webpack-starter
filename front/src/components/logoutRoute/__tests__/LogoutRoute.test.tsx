import React from 'react';
import { shallow } from 'enzyme';
import { MemoryRouter } from 'react-router';
import { AuthProvider } from '../../../contexts/auth';
import LogoutRoute from '../index';

describe('LogoutRoute component', () => {
  it('renders as expected', () => {
    const props = {};

    const component = shallow(
      <MemoryRouter initialEntries={['/']}>
        <AuthProvider>
          <LogoutRoute {...props} />
        </AuthProvider>
      </MemoryRouter>,
    );
    expect(component).toMatchSnapshot();
  });
});
