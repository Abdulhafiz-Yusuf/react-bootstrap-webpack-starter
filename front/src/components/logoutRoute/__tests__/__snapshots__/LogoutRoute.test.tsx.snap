// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogoutRoute component renders as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MemoryRouter
    initialEntries={
      Array [
        "/",
      ]
    }
  >
    <AuthProvider
      initialState={
        Object {
          "isAuthenticated": false,
          "isExpiredToken": true,
          "lastAuthDate": undefined,
          "token": "",
          "user": undefined,
        }
      }
    >
      <LogoutRoute />
    </AuthProvider>
  </MemoryRouter>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <AuthProvider
        initialState={
          Object {
            "isAuthenticated": false,
            "isExpiredToken": true,
            "lastAuthDate": undefined,
            "token": "",
            "user": undefined,
          }
        }
      >
        <LogoutRoute />
      </AuthProvider>,
      "history": Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "wfg6nz",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "wfg6nz",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <LogoutRoute />,
        "initialState": Object {
          "isAuthenticated": false,
          "isExpiredToken": true,
          "lastAuthDate": undefined,
          "token": "",
          "user": undefined,
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <AuthProvider
          initialState={
            Object {
              "isAuthenticated": false,
              "isExpiredToken": true,
              "lastAuthDate": undefined,
              "token": "",
              "user": undefined,
            }
          }
        >
          <LogoutRoute />
        </AuthProvider>,
        "history": Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "key": "wfg6nz",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "wfg6nz",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <LogoutRoute />,
          "initialState": Object {
            "isAuthenticated": false,
            "isExpiredToken": true,
            "lastAuthDate": undefined,
            "token": "",
            "user": undefined,
          },
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
